plugin/pulpcore/plugin/stages/content_unit_stages.py

```diff
@@ -52,7 +52,8 @@ class QueryExistingContentUnits(Stage):
 
             for model_type in content_q_by_type.keys():
                 for result in model_type.objects.filter(content_q_by_type[model_type]):
-                    for declarative_content in batch:
+                    model_batch = filter(lambda dcont: type(dcont.content) == model_type, batch)
```

Comment:
I only provided the `batch` iterator for this code, so I am not the expert here.

These nested loops look like they have the word "SLOW" imprinted on them (but let's wait for the outcome of the performance testing).

Given the current structure, it looks like the right thing to do. However, I would prefer staying imperative style here. There is already a `continue` in this loop, so I would prefer:

```python
                    for declarative_content in batch:
                        if type(dcont.content) is not model_type:
                            continue
                        not_same_unit = False
...
```
 An alternative would be to sort the batch into buckets by content type in the loop above (analogous to how the query is built) and iterate through the specific bucket here.